app/src/main/java/com/example/tawazz/storage/Storage.java

```diff
@@ -0,0 +1,54 @@
+package com.example.tawazz.storage;
+
+import android.net.Uri;
+
+import com.example.tawazz.download.DownloadInvoker;
+import com.example.tawazz.download.ExtensionType;
+import com.example.tawazz.storage.exceptions.FailedStoreException;
+import com.example.tawazz.task.TaskSuccessfulWaiterFactory;
+import com.example.tawazz.utils.FailedWaitingForCondition;
+import com.google.android.gms.tasks.OnSuccessListener;
+import com.google.firebase.storage.StorageReference;
+import com.google.firebase.storage.UploadTask;
+
+public class Storage {
+    private StorageReference mStorageReference;
+    private TaskSuccessfulWaiterFactory mTaskSuccessfulWaiterFactory;
+    private DownloadInvoker mDownloadInvoker;
+
+    public Storage(StorageReference storageReference, TaskSuccessfulWaiterFactory taskSuccessfulWaiterFactory, DownloadInvoker mDownloadInvoker) {
+        this.mStorageReference = storageReference;
+        this.mTaskSuccessfulWaiterFactory = taskSuccessfulWaiterFactory;
+        this.mDownloadInvoker = mDownloadInvoker;
+    }
+
+    public UploadTask store(Uri srcDir, Uri destDir) {
+        StorageReference userIconStorageRef = mStorageReference.child(destDir.getPath());
+        return userIconStorageRef.putFile(srcDir);
+    }
+
+    public void safeStore(Uri srcDir, Uri destDir) throws FailedStoreException {
```

Comment:
this is not just a safeStore, it also waits unlike the `store` method. consider renaming those 2 methods